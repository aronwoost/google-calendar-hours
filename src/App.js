import React from 'react';

const App = () => <span>hello</span>;

export default App;
