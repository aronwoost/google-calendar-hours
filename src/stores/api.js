import queryString from 'query-string';

const API_BASE_PATH = 'https://www.googleapis.com/calendar/v3/';

const fetchGoogle = ({ url, accessToken, params }) =>
  fetch(
    `${url}?${queryString.stringify({
      access_token: accessToken,
      ...params,
    })}`
  ).then((response) => {
    if (response.status !== 200) {
      localStorage.removeItem('accessToken');
      window.location = '/';
    }
    return response.json();
  });

export const fetchCalendars = ({ accessToken }) =>
  fetchGoogle({ url: `${API_BASE_PATH}users/me/calendarList`, accessToken });

export const fetchCalendarEvents = ({ accessToken, calendarId }) =>
  fetchGoogle({
    url: `${API_BASE_PATH}calendars/${calendarId}/events`,
    accessToken,
    params: {
      singleEvents: true,
      maxResults: 2500,
    },
  });
