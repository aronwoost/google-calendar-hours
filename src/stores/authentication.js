import { createSlice } from '@reduxjs/toolkit';

export const authentication = createSlice({
  name: 'authentication',
  initialState: null,
  reducers: {},
});

export const selectAccessToken = (state) => state.authentication.accessToken;
export const selectHasToken = (state) => !!selectAccessToken(state);

export default authentication.reducer;
