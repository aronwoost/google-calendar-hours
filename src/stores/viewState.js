import { createSlice } from '@reduxjs/toolkit';
import dayjs from 'dayjs';
import 'dayjs/locale/de';

import { loadCalendarEvents, selectCalendarEvents } from './calendarEvents';
import { updateConfig } from './storage';
import { RANGE_TYPE, WEEK_START } from '../constants';
import roundHours from '../utils/roundHours';

export const viewState = createSlice({
  name: 'viewState',
  initialState: null,
  reducers: {
    setSelectedCalendarId: (state, { payload }) => {
      state.selectedCalendarId = payload;
    },
    setRangeType: (state, { payload }) => {
      state.selectedRangeType = payload;
    },
    changeRange: (state, { payload }) => {
      if (payload === 'prev') {
        state.currentDatePointerStart = dayjs(state.currentDatePointerStart)
          .subtract(1, state.selectedRangeType)
          .toJSON();
      } else if (payload === 'next') {
        state.currentDatePointerStart = dayjs(state.currentDatePointerStart)
          .add(1, state.selectedRangeType)
          .toJSON();
      }
    },
    resetRange: (state) => {
      state.currentDatePointerStart = dayjs().startOf('day').toJSON();
    },
    setWeekStart: (state, { payload }) => {
      state.weekStart = payload;
    },
    setStart: (state, { payload }) => {
      state.currentDatePointerStart = payload;
    },
    setEnd: (state, { payload }) => {
      state.currentDatePointerEnd = payload;
    },
  },
});

export const { changeRange, resetRange } = viewState.actions;
const {
  setSelectedCalendarId,
  setRangeType,
  setWeekStart,
  setStart,
  setEnd,
} = viewState.actions;

export const selectSelectedCalendar = (state) =>
  state.viewState.selectedCalendarId;

export const selectDate = (state) => state.viewState.currentDatePointerStart;

export const selectRangeType = (state) => state.viewState.selectedRangeType;
export const selectWeekStart = (state) => state.viewState.weekStart;
export const selectLocaleForWeekStart = (state) =>
  state.viewState.weekStart === WEEK_START.SUNDAY ? 'en' : 'de';

export const selectCurrentDatePointers = (state) => {
  const {
    selectedRangeType,
    currentDatePointerStart,
    currentDatePointerEnd,
  } = state.viewState;
  const currentDatePointerStartDate = dayjs(currentDatePointerStart);

  if (selectedRangeType === RANGE_TYPE.CUSTOM) {
    return {
      start: dayjs(currentDatePointerStart),
      // The selected end day should be included in the calculation, so we
      // need to add an extra day.
      end: dayjs(currentDatePointerEnd).add(1, 'day'),
    };
  }

  let rangeStart;
  let rangeEnd;

  if (selectedRangeType === RANGE_TYPE.DAY) {
    rangeStart = currentDatePointerStartDate.startOf('day');
    rangeEnd = rangeStart.add(1, 'day');
  } else if (selectedRangeType === RANGE_TYPE.WEEK) {
    rangeStart = currentDatePointerStartDate
      .locale(selectLocaleForWeekStart(state))
      .startOf('day')
      .weekday(0);
    rangeEnd = rangeStart.add(1, 'week');
  } else if (selectedRangeType === RANGE_TYPE.MONTH) {
    rangeStart = currentDatePointerStartDate.startOf('month');
    rangeEnd = rangeStart.add(1, 'month');
  } else if (selectedRangeType === RANGE_TYPE.YEAR) {
    rangeStart = currentDatePointerStartDate.startOf('year');
    rangeEnd = rangeStart.add(1, 'year');
  } else if (selectedRangeType === RANGE_TYPE.TOTAL) {
    rangeStart = dayjs('2000-01-01T10:00:00Z');
    rangeEnd = dayjs('2040-01-01T10:00:00Z');
  }

  return {
    start: rangeStart,
    end: rangeEnd,
  };
};

export const selectEventsByRange = (state) => {
  const events = selectCalendarEvents(state, selectSelectedCalendar(state));

  if (!events) {
    return null;
  }

  const { start: rangeStart, end: rangeEnd } = selectCurrentDatePointers(state);

  return events
    .filter(
      ({ start, end }) =>
        // Filter out all events that end before selected start date or start after
        // selected end date.
        !(new Date(end) < rangeStart || new Date(start) > rangeEnd)
    )
    .map(({ start, end, ...rest }) => ({
      ...rest,
      start: new Date(start) < rangeStart ? rangeStart.toJSON() : start,
      end: new Date(end) > rangeEnd ? rangeEnd.toJSON() : end,
    }));
};

export const selectHours = (state) => {
  const events = selectEventsByRange(state);

  if (!events) {
    return null;
  }

  let hours = 0;

  events.forEach(({ start, end }) => {
    hours += (new Date(end) - new Date(start)) / 1000 / 60 / 60;
  });

  return roundHours(hours);
};

export const setSelectedCalendar = ({ calendarId }) => (dispatch, getState) => {
  dispatch(setSelectedCalendarId(calendarId));
  updateConfig({ selectedCalendarId: calendarId });
  const calendarEvents = selectCalendarEvents(getState(), calendarId);
  if (!calendarEvents) {
    dispatch(loadCalendarEvents({ calendarId }));
  }
};

export const changeRangeType = ({ range }) => (dispatch, getState) => {
  if (range === RANGE_TYPE.CUSTOM) {
    const { start, end } = selectCurrentDatePointers(getState());
    // We need to subtract a day here, because a day was added to `end`
    // in order to have the selected end day in the calculation.
    const correctedEnd = end.subtract(1, 'day');
    dispatch(setStart(start.toJSON()));
    dispatch(setEnd(correctedEnd.toJSON()));
    updateConfig({ start: start.toJSON(), end: correctedEnd.toJSON() });
  }
  dispatch(setRangeType(range));
  updateConfig({ selectedRangeType: range });
};

export const changeWeekStart = (weekStart) => (dispatch) => {
  dispatch(setWeekStart(weekStart));
  updateConfig({ weekStart });
};

export const changeStart = (start) => (dispatch) => {
  dispatch(setStart(start));
  updateConfig({ start });
};

export const changeEnd = (end) => (dispatch) => {
  dispatch(setEnd(end));
  updateConfig({ end });
};

export default viewState.reducer;
