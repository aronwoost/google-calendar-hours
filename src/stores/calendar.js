import { createSlice } from '@reduxjs/toolkit';
import { get } from 'lodash';

import { selectAccessToken } from './authentication';
import { fetchCalendarEvents } from './api';

export const calendar = createSlice({
  name: 'calendar',
  initialState: {
    map: {},
  },
  reducers: {
    setCalendarEvents: (state, { payload }) => {
      state.map[payload.calendarId] = payload.events;
    },
  },
});

const { setCalendarEvents } = calendar.actions;

export const loadCalendarEvents = ({ calendarId }) => async (
  dispatch,
  getState
) => {
  const state = getState();
  const accessToken = selectAccessToken(state);
  try {
    const data = await fetchCalendarEvents({ accessToken, calendarId });
    dispatch(setCalendarEvents({ calendarId, events: data.items }));
    return Promise.resolve();
  } catch (e) {
    return Promise.reject();
  }
};

export const selectCalendarEvents = (state, calendarId) =>
  get(state, `calendar.map["${calendarId}"]`, null);

export default calendar.reducer;
