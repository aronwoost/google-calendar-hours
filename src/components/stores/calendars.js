import { createSlice } from '@reduxjs/toolkit';

import { selectAccessToken } from './authentication';
import { fetchCalendars } from './api';
import { setSelectedCalendar } from './viewState';
import { getConfig } from './storage';

export const calendars = createSlice({
  name: 'calendars',
  initialState: {
    list: null,
  },
  reducers: {
    setCalendars: (state, { payload }) => {
      state.list = payload;
    },
  },
});

const { setCalendars } = calendars.actions;

export const loadCalendars = () => async (dispatch, getState) => {
  const accessToken = selectAccessToken(getState());
  try {
    const { items } = await fetchCalendars({ accessToken });
    const calendarList = items.map(({ id, summary }) => ({
      id,
      label: summary,
    }));

    dispatch(setCalendars(calendarList));

    const { selectedCalendarId } = getConfig() ?? {};

    if (selectedCalendarId) {
      const calendarExists = calendarList?.find(
        ({ id }) => id === selectedCalendarId
      );

      if (calendarExists) {
        dispatch(setSelectedCalendar({ calendarId: selectedCalendarId }));
      } else {
        // dispatch null, so that it's removed from localStorage config
        dispatch(setSelectedCalendar({ calendarId: null }));
      }
    }
  } catch (e) {
    // do nothing
  }
};

export const selectCalendars = (state) => state.calendars.list;

export default calendars.reducer;
