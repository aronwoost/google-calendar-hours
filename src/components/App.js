'use client';

import React, { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { decode } from 'qss';

import { selectHasToken } from './stores/authentication';

import AuthScreen from './AuthScreen';
import Interface from './Interface';

import styles from './App.module.css';

const App = () => {
  const hasToken = useSelector(selectHasToken);
  const [isClient, setIsClient] = React.useState(false);

  useEffect(() => {
    const hashParams = decode(window.location?.hash?.slice(1) ?? '');

    if (hashParams.access_token) {
      sessionStorage.setItem('accessToken', hashParams.access_token);
      window.location = '/';
    }
  }, []);

  useEffect(() => {
    setIsClient(true);
  }, []);

  return (
    <div className={styles.appWrapper}>
      <div className={styles.app}>
        <div className={styles.sticky}>
          <div className={styles.content}>
            <h1 className={styles.headline}>
              Google Calendar Hours Calculator
            </h1>
            {(!isClient || !hasToken) && <AuthScreen />}
            {isClient && hasToken && <Interface />}
          </div>
          <footer className={styles.footer}>
            <p>
              <span>Â© 2011 - 2024. This app is open source. </span>
              <a
                href="https://github.com/aronwoost/google-calendar-hours"
                target="_blank"
                rel="noreferrer"
              >
                Check it on GitHub
              </a>
              <span>.</span>
            </p>
          </footer>
        </div>
      </div>
    </div>
  );
};

export default App;
