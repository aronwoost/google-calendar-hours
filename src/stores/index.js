import { configureStore } from '@reduxjs/toolkit';
import dayjs from 'dayjs';
import weekday from 'dayjs/plugin/weekday';

import authentication from './authentication';
import calendars from './calendars';
import viewState from './viewState';
import calendarEvents from './calendarEvents';
import { getConfig } from './storage';
import { RANGE_TYPE, WEEK_START } from '../constants';

dayjs.extend(weekday);

const getAccessToken = () => {
  try {
    return sessionStorage.getItem('accessToken');
  } catch (e) {
    // don't handle
  }

  return null;
};

const getViewState = () => {
  const config = getConfig() ?? {};
  return {
    selectedRangeType: config.selectedRangeType ?? RANGE_TYPE.MONTH,
    currentDatePointerStart:
      config.selectedRangeType === RANGE_TYPE.CUSTOM
        ? config.start
        : dayjs().startOf('day').toJSON(),
    currentDatePointerEnd: config.end,
    weekStart: config.weekStart ?? WEEK_START.MONDAY,
  };
};

const store = () =>
  configureStore({
    reducer: {
      authentication,
      calendars,
      viewState,
      calendarEvents,
    },
    preloadedState: {
      authentication: { accessToken: getAccessToken() },
      viewState: getViewState(),
    },
  });

export default store;
