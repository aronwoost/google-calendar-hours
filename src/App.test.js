import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import timekeeper from 'timekeeper';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

import { store } from './stores';
import App from './App';
import { getInitialState } from './stores/viewState';

const createTestStore = ({
  authentication,
  viewState,
  calendars,
  calendarEvents,
}) =>
  store({
    initialState: {
      authentication: { accessToken: 'ABC123', ...authentication },
      viewState: {
        ...getInitialState(),
        selectedCalendarId: 'test-id',
        ...viewState,
      },
      calendars: {
        list: [{ id: 'test-id', label: 'test-name' }],
        ...calendars,
      },
      calendarEvents,
    },
  });

const renderAppWithStore = ({
  authentication,
  viewState,
  calendars,
  calendarEvents,
} = {}) =>
  render(
    <Provider
      store={createTestStore({
        authentication,
        viewState,
        calendars,
        calendarEvents,
      })}
    >
      <App />
    </Provider>
  );

beforeEach(() => {
  timekeeper.freeze(new Date('2018-01-01T10:00:00Z'));
  window.localStorage.removeItem('config');
});

const testEvents = [
  {
    start: { dateTime: '2016-01-01T10:00:00Z' },
    end: { dateTime: '2016-01-01T11:00:00Z' },
  },
  {
    start: { dateTime: '2018-01-02T10:00:00Z' },
    end: { dateTime: '2018-01-02T11:00:00Z' },
  },
];

const server = setupServer(
  rest.get(
    'https://www.googleapis.com/calendar/v3/users/me/calendarList',
    (req, res, ctx) => {
      const accessToken = req.url.searchParams.get('access_token');
      if (accessToken !== 'ABC123') {
        return res(403);
      }

      return res(
        ctx.json({ items: [{ id: 'test-id', summary: 'test-name' }] })
      );
    }
  ),
  rest.get(
    'https://www.googleapis.com/calendar/v3/calendars/test-id/events?maxResults=2500&singleEvents=true',
    (req, res, ctx) => {
      const accessToken = req.url.searchParams.get('access_token');
      if (accessToken !== 'ABC123') {
        return res(403);
      }

      return res(ctx.json({ items: testEvents }));
    }
  )
);

delete window.location;

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

it('renders google auth button', () => {
  const { getByAltText } = renderAppWithStore({
    authentication: { accessToken: null },
  });

  expect(getByAltText('Auth with Google')).toBeInTheDocument();
});

it('writes access token to localStorage and does redirect', () => {
  window.location = new URL(
    'https://www.example.com/hello#access_token=ABC123'
  );

  renderAppWithStore({
    authentication: { accessToken: null },
  });

  expect(window.localStorage.getItem('accessToken')).toEqual('ABC123');
  expect(window.location).toBe('/');
});

it('renders "loading" without calendars', () => {
  const { getByText } = renderAppWithStore({ calendars: { list: null } });

  expect(getByText('loading')).toBeInTheDocument();
});

it('renders calendars list', () => {
  const { queryByText } = renderAppWithStore();

  expect(queryByText('Please select calendar')).not.toBeInTheDocument();
});

it('requests calendars and display placeholder', async () => {
  const { findByText, queryByTestId, queryByText } = renderAppWithStore({
    viewState: { selectedCalendarId: null },
    calendars: { list: null },
  });

  expect(await findByText('Please select calendar')).toBeInTheDocument();
  expect(await findByText('test-name')).toBeInTheDocument();
  expect(queryByTestId('RangeSelectList')).not.toBeInTheDocument();
  expect(queryByText('Loading hours')).not.toBeInTheDocument();
  expect(queryByTestId('RangeChanger')).not.toBeInTheDocument();
});

it('renders range select list and hours (happy path)', () => {
  const { getByText } = renderAppWithStore({
    calendarEvents: {
      map: { 'test-id': testEvents },
    },
  });

  expect(getByText('Total')).toBeInTheDocument();
  expect(getByText('Week')).toBeInTheDocument();
  expect(getByText('2h')).toBeInTheDocument();
});

it('renders correctly after user changes calendar', () => {
  const { getByTestId, getByText } = renderAppWithStore({
    calendars: {
      list: [
        { id: 'test-id', label: 'test-name' },
        { id: 'test-id-2', label: 'test-name-2' },
      ],
    },
    calendarEvents: {
      map: { 'test-id': [], 'test-id-2': testEvents },
    },
  });

  fireEvent.change(getByTestId('CalendarsList'), {
    target: { value: 'test-id-2' },
  });

  expect(getByText('2h')).toBeInTheDocument();
});

it('requests events, display hours and sets localStorage when loaded', async () => {
  const { getByTestId, findByText } = renderAppWithStore();

  fireEvent.change(getByTestId('CalendarsList'), {
    target: { value: 'test-id' },
  });

  expect(await findByText('2h')).toBeInTheDocument();

  expect(window.localStorage.getItem('config')).toEqual(
    '{"selectedCalendarId":"test-id"}'
  );
});

describe('localStorage', () => {
  it('saves user selection to localStorage', () => {
    const { getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: { 'test-id': testEvents },
      },
    });

    fireEvent.change(getByTestId('CalendarsList'), {
      target: { value: 'test-id' },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'week' },
    });

    expect(window.localStorage.getItem('config')).toEqual(
      '{"selectedCalendarId":"test-id","selectedRangeType":"week"}'
    );
  });

  it('uses data from localStorage to sets UI', async () => {
    window.localStorage.setItem(
      'config',
      JSON.stringify({
        selectedCalendarId: 'test-id',
        selectedRangeType: 'week',
      })
    );

    const { findByText } = renderAppWithStore({
      viewState: {
        selectedCalendarId: null,
      },
      calendars: {
        list: null,
      },
    });

    expect(await findByText('1h')).toBeInTheDocument();
  });
});

describe('calculate hours', () => {
  it('renders 0h hours (if not matching events)', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-20T10:00:00Z' },
              end: { dateTime: '2018-01-20T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'week' },
    });

    expect(getByText('0h')).toBeInTheDocument();
  });

  it('renders hours for day', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-10T10:00:00Z' },
              end: { dateTime: '2018-01-10T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'day' },
    });

    expect(getByText('2h')).toBeInTheDocument();
  });

  it('renders hours for day when user changes to previous day', () => {
    timekeeper.freeze(new Date('2018-01-02T10:00:00Z'));

    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-02T10:00:00Z' },
              end: { dateTime: '2018-01-02T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'day' },
    });

    fireEvent.click(getByText('Prev'));

    expect(getByText('2h')).toBeInTheDocument();
  });

  it('renders hours for day when user changes to next day', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-02T10:00:00Z' },
              end: { dateTime: '2018-01-02T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'day' },
    });

    fireEvent.click(getByText('Next'));

    expect(getByText('1h')).toBeInTheDocument();
  });

  it('renders hours for day when user resets', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-02T10:00:00Z' },
              end: { dateTime: '2018-01-02T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'day' },
    });

    fireEvent.click(getByText('Next'));

    fireEvent.click(getByText('Reset'));

    expect(getByText('2h')).toBeInTheDocument();
  });

  it('renders hours for week', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-02T10:00:00Z' },
              end: { dateTime: '2018-01-02T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-10T10:00:00Z' },
              end: { dateTime: '2018-01-10T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'week' },
    });

    expect(getByText('2h')).toBeInTheDocument();
  });

  it('renders hours for week when user changes to previous week', () => {
    timekeeper.freeze(new Date('2018-01-12T10:00:00Z'));

    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-12T10:00:00Z' },
              end: { dateTime: '2018-01-12T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'week' },
    });

    fireEvent.click(getByText('Prev'));

    expect(getByText('2h')).toBeInTheDocument();
  });

  it('renders hours for week when user changes to next week', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-12T10:00:00Z' },
              end: { dateTime: '2018-01-12T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'week' },
    });

    fireEvent.click(getByText('Next'));

    expect(getByText('1h')).toBeInTheDocument();
  });

  it('renders hours for week when user resets', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-12T10:00:00Z' },
              end: { dateTime: '2018-01-12T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'week' },
    });

    fireEvent.click(getByText('Next'));

    fireEvent.click(getByText('Reset'));

    expect(getByText('2h')).toBeInTheDocument();
  });

  it('renders hours for month', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-14T13:00:00Z' },
              end: { dateTime: '2018-01-14T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-02-01T10:00:00Z' },
              end: { dateTime: '2018-02-01T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'month' },
    });

    expect(getByText('2h')).toBeInTheDocument();
  });

  it('renders hours for month when user changes to previous month', () => {
    timekeeper.freeze(new Date('2018-02-01T10:00:00Z'));

    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-02-01T10:00:00Z' },
              end: { dateTime: '2018-02-01T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'month' },
    });

    fireEvent.click(getByText('Prev'));

    expect(getByText('2h')).toBeInTheDocument();
  });

  it('renders hours for month when user changes to next month', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-02-01T10:00:00Z' },
              end: { dateTime: '2018-02-01T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'month' },
    });

    fireEvent.click(getByText('Next'));

    expect(getByText('1h')).toBeInTheDocument();
  });

  it('renders hours for month when user resets', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-02-01T10:00:00Z' },
              end: { dateTime: '2018-02-01T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'month' },
    });

    fireEvent.click(getByText('Next'));

    fireEvent.click(getByText('Reset'));

    expect(getByText('2h')).toBeInTheDocument();
  });

  it('renders hours for year', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2017-01-01T10:00:00Z' },
              end: { dateTime: '2017-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-14T13:00:00Z' },
              end: { dateTime: '2018-01-14T14:00:00Z' },
            },
            {
              start: { dateTime: '2018-02-01T10:00:00Z' },
              end: { dateTime: '2018-02-01T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'year' },
    });

    expect(getByText('2h')).toBeInTheDocument();
  });

  it('renders hours for year when user changes to previous year', () => {
    timekeeper.freeze(new Date('2019-01-01T10:00:00Z'));

    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2019-01-01T10:00:00Z' },
              end: { dateTime: '2019-01-01T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'year' },
    });

    fireEvent.click(getByText('Prev'));

    expect(getByText('2h')).toBeInTheDocument();
  });

  it('renders hours for year when user changes to next year', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2019-01-01T10:00:00Z' },
              end: { dateTime: '2019-01-01T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'year' },
    });

    fireEvent.click(getByText('Next'));

    expect(getByText('1h')).toBeInTheDocument();
  });

  it('renders hours for year when user resets', () => {
    const { getByText, getByTestId } = renderAppWithStore({
      calendarEvents: {
        map: {
          'test-id': [
            {
              start: { dateTime: '2018-01-01T10:00:00Z' },
              end: { dateTime: '2018-01-01T11:00:00Z' },
            },
            {
              start: { dateTime: '2018-01-01T13:00:00Z' },
              end: { dateTime: '2018-01-01T14:00:00Z' },
            },
            {
              start: { dateTime: '2019-01-01T10:00:00Z' },
              end: { dateTime: '2019-01-01T11:00:00Z' },
            },
          ],
        },
      },
    });

    fireEvent.change(getByTestId('RangeSelectList'), {
      target: { value: 'year' },
    });

    fireEvent.click(getByText('Next'));

    fireEvent.click(getByText('Reset'));

    expect(getByText('2h')).toBeInTheDocument();
  });
});

describe('display time range in human readable format', () => {
  it('renders current day', () => {
    const { getByText } = renderAppWithStore({
      viewState: { selectedRangeType: 'day' },
      calendarEvents: { map: { 'test-id': [] } },
    });

    expect(getByText('Monday, January 1, 2018')).toBeInTheDocument();
  });

  it('renders current week', () => {
    const { getByText } = renderAppWithStore({
      viewState: { selectedRangeType: 'week' },
      calendarEvents: { map: { 'test-id': [] } },
    });

    expect(getByText('01.01.2018 - 08.01.2018')).toBeInTheDocument();
  });

  it('renders current month', () => {
    const { getByText } = renderAppWithStore({
      viewState: { selectedRangeType: 'month' },
      calendarEvents: { map: { 'test-id': [] } },
    });

    expect(getByText('January, 2018')).toBeInTheDocument();
  });

  it('renders current year', () => {
    const { getByText } = renderAppWithStore({
      viewState: { selectedRangeType: 'year' },
      calendarEvents: { map: { 'test-id': [] } },
    });

    expect(getByText('2018')).toBeInTheDocument();
  });

  it('renders without RangeDisplay ("total")', () => {
    const { queryByTestId } = renderAppWithStore({
      viewState: { selectedRangeType: 'total' },
      calendarEvents: { map: { 'test-id': [] } },
    });

    expect(queryByTestId('RangeDisplay')).not.toBeInTheDocument();
  });
});
