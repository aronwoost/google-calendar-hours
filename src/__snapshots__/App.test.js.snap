// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display events renders events 1`] = `
"Start,End,Title,Hours
2018-01-01T10:00:00.000Z,2018-01-01T12:00:00.000Z,\\"event-1\\",2
2018-01-05T13:00:00.000Z,2018-01-05T18:00:00.000Z,\\"event-2\\",5"
`;
